Differential (Additive/Removal) patching when switching between design / debug

apisaint opened the pull request:
😩 Finally got tired of dealing with it so I tried fixing it.

This patch adds differential patching for imports that occur during the switching between design and debug tabs inside of Designer. As reported through #2971, and others, this patch favors existing data over imported data, except urls (see below), values that exist already on the document will remain unchanged, only new values (including array based values) will be added / removed.

This also includes a bypass feature for urls, currently these options are not exposed through the interface but could be. This feature is an object to allow for future properties for preference-based patching.

This also allows endpoints to stay in position when moved or grouped in folders.

👇 Included:

- Adds `options.enableDiffBasedPatching` and `options.enableDiffDeep` to `importRaw` for backwards compat.
- Adds `options.bypassDiffProps.url` for url bypassing and use an object for future items.
- Adds `diffPatchObj` based on differential patching for objects (works with arrays as well).

💭 Future ideas:

- `hasBeenModifiedByUser` property map object to allow changing properties that haven't been touched by the user with options.
- Notice icon when clicked displays dialogue with changes that a user can choose to apply. Underlying diff method can be augmented to create such patch objects or use a lib.

⚠️ Notes:

Basically this makes changes additive, which is better than the current scenario especially when paired with #2979, what does that imply? See below.

- If it exists it won’t change unless a bypass option is created.
- When the `url` changes for a Request, when the bypass is enabled (true by default here) it will update the url in the url bar but it does not change the `name`, so the sidebar will not auto-update. This can be resolved in a future or with a modified check for the name using the future idea explained above.
- When items / values inside of the Headers change it does not auto-update, this would require either a bypass option, or rely on the future idea.

Related #3038
Closes #2971, #2882, #2442

reynolek commented:
We have this on our list of issues to tackle, thanks for taking the time to look into it! Going to add @nijikokun to review it from a product perspective along with @develohpanda and @dimitropoulos for code.

nijikokun commented on packages/insomnia-app/app/common/import.js:
```diff
@@ -240,7 +241,19 @@ export async function importRaw(
+      let updateDoc = resource;
+
+      // Do differential patching when enabled
+      if (options.enableDiffBasedPatching) {
```
`options` isn't checked for existence so this will fail with other import methods.

apisaint replied:
Good catch, instead of worrying about existence decided to default the argument to an object and let type checking do the worrying about input for us.

nijikokun commented:
Oh wow. Taking a look now  👀 

> * backwards compat.

Beautiful 🙌 

> This also allows endpoints to stay in position when moved or grouped in folders.

Absolutely amazing 🎉 

> * it will update the url in the url bar but it does not change the `name`, so the sidebar will not auto-update. This can be resolved in a future or with a modified check for the name using the future idea explained above.

While we do have `modified` you're right we don't have something that tells us whether it's been modified by the user or through creation, at least to my knowledge, so I agree this can be a future improvement. 👍 

> * Notice icon when clicked displays dialogue with changes that a user can choose to apply. Underlying diff method can be augmented to create such patch objects or use a lib.

Love this idea 💯, maybe this could also be expanded to sync at some point 🤔

> * `hasBeenModifiedByUser` property map object to allow changing properties that haven't been touched by the user with options.

Makes sense, track properties that change, and update the ones that haven't, seems like a good future improvement.  💯 

> * Adds `options.bypassDiffProps.url` for url bypassing and use an object for future items.

👍 

---

Amazing, love the breakdown. Found an issue in the code, but other than that it's definitely an improvement over the current situation and I'll take it.

dimitropoulos committed:
adds some basic tests

- also adds jest (which was (mistakenly) not there before)
- does not call `.hasOwnProperty` directly, per https://eslint.org/docs/rules/no-prototype-builtins (which, we will add more formally at a later date)

dimitropoulos commented:
(working on making sure I correctly resolve the merge conflict of the prior commit on master then this is next up to be merged)

dimitropoulos committed:
don't special-case workspaces

this ensures the behavior of the initial PR is more preserved

dimitropoulos committed:
fix(import) do differential patching for design activity imports

This patch adds differential patching for imports that occur during the switching between
design and debug tabs inside of Designer. As reported through #2971, and others, this patch
favors existing data over imported data, values that exist already on the document will remain
unchanged, only new values (including array based values) will be added / removed.

This also includes a bypass feature for urls, currently these options are not exposed through
the interface but could be. This feature is an object to allow for future properties for
preference-based patching.

- Adds `options.enableDiffBasedPatching` and `options.enableDiffDeep` to `importRaw` for backwards compat.
- Adds `options.bypassDiffProps.url` for url bypassing and use an object for future items.
- Adds `diffPatchObj` based on differential patching for objects (works with arrays as well).

Future ideas:

- `hasBeenModifiedByUser` property map object to allow changing properties that haven't been touched by the user with options.

fixes: #2971, #2882, #3038, #2442
